### ISSUE TITLE
EDB Language Pack for Windows Discontinued: Configuring Python Support

### ISSUE THREAD

**Issue by anthonydb** [anthonydb/practical-sql-2#32]
Unfortunately, as of PostgreSQL 17, [the Windows download from EDB](https://github.com/EnterpriseDB/edb-installers/issues/184) described on page 5 of Chapter 1 no longer includes the Language Pack that had been part of the Stack Builder utility. 

Instead, to configure Python for the exercise that uses a Python function in Chapter 17, please follow these revised steps:

1. PostgreSQL 17 for Windows currently requires Python 3.12.6. Download the official installer at  [https://www.python.org/ftp/python/3.12.6/python-3.12.6-amd64.exe](https://www.python.org/ftp/python/3.12.6/python-3.12.6-amd64.exe)
2. Performing the custom installation and choose a directoy that's easy to remember. For example, `C:\Python`
3. Next, follow the directions on Page 6 of Chapter 1 to add that directory to the `PATH` and `PYTHONHOME` variables under `System Variables` Instead of adding the directory shown in the book, add the directory where you installed Python
4. Restart your PC for the changes to take effect.

You now should be able to create the `PL/Python` extension in a database by running:

`CREATE EXTENSION plpython3u;`

**Note**

You can find additional information in the directory on your PC where PostgreSQL is installed, typically `C:\Program File\PostgreSQL\17\doc\installation-notes.html`

**Comment by rafissif** [comment]
Thank you so much, sir...

**Comment by EmunahLife5785** [comment]
Thank you for taking the time post this update!

**Comment by marcodalo** [comment]
Thank you for this post. It is a shame that EDB does not make it clear before downloading that their Pg17 / 18 installers no longer support plpython3u and the other languages that according to the PostgreSQL Development group are part of the standard (EDB) installation. 
However, I followed your advice, but I could not get the plpython3u extension to work. Where do I find the Chapter 1 page 5/6 that your are referring to?

**Comment by anthonydb** [comment]
@marcodalo Those instructions are in Chapter 1, starting on page 5, of the 2nd edition of "Practical SQL." Which version are you reading?

**Comment by marcodalo** [comment]
I was reading the 2nd edition, and nothing in there was new to me. Yet the plpythpn3u extension does not install properly. I am working with PostgreSQL 18.2, Windows 10 fully updated and Python 3.13 (as required by Pg 18). If you can manage, I, and many others will be quite interested to see your solution. EDB has announced that they no longer support plpython on Windows. The announcement is hidden in the Pg18.2 installation notes - very user friendly. I am not sure where that leaves plpython users.

**Comment by anthonydb** [comment]
@marcodalo Did you follow the steps outlined above in this announcement?

**Comment by marcodalo** [comment]
Yes, and I tried variations with and without the EDB language pack. I spent a full day researching the topic and trying various solution ideas, but nothing worked. I am a long-time user of plpython, and I have a stack of Postgres python functions. Which is why I am not pleased with EDB scrapping its plpython support.

**Comment by 197788** [comment]
> I am not pleased with EDB scrapping its plpython support.

...time to switch fully to Linux... did that years ago. check out [awesome-postgres](https://github.com/dhamaniasad/awesome-postgres) You may find something else that does work for you.

**Comment by marcodalo** [comment]
I checked awesome-postgres but I could not find a solution. Have you seen one? 
My conclusion is the same as yours: get the hell out of Windows. That is really the message that EDB and, indirectly, the PostgreSQL Development group, is sending to users. They are free to do so, but EDB should, at a very minimum, have the decency to make this clear before users start installing their packages and subsequently waste hours or days trying to figure out what went wrong. Sure, there is a notice in the installation notes... which appear AFTER the installation.

**Comment by anthonydb** [comment]
@marcodalo I share your frustration. Keeping the readers of Practical SQL up to date with changes EDB has made to its PostgreSQL installer for Windows (and to StackBuilder and pgAdmin) has been a challenge. In regards to Python, I have been able to successfully run PL/Python by following the steps at the top of this issue. The only way it works is if you have the exact supported version of Python for the version of PostgreSQL installed AND if you set the to the `PATH` and `PYTHONHOME` variables to include the directory where Python is installed. Restart the machine, and it should work.

**Comment by marcodalo** [comment]
I have Windows 10, PostgreSQL 18.2, Python 3.13, PYTHOHHOME = "C:\Users\Marc\AppData\Local\Programs\Python\Python313", PYTHONPATH = "C:\Users\Marc\AppData\Local\Programs\Python\Python313" in System Env Var and the same in PATH both in System and User Env Var. Perhaps PYTHONPATH is an overkill, but I have tried with and without.
C:/Program Files/PostgreSQL/18/lib/plpython3.dll is present - can be seen in File Explorer.

CREATE EXTENSION plpython3u;
I get
[58P01] ERROR: could not load library "C:/Program Files/PostgreSQL/18/lib/plpython3.dll": The specified module could not be found.

  Where: SQL statement "CREATE FUNCTION plpython3_call_handler() RETURNS language_handler
  LANGUAGE c AS '$libdir/plpython3'"
extension script file "plpython3u--1.0.sql", near line 3

Any idea?

**Comment by 197788** [comment]
> Any idea?

Maybe you can make another user account on your Windows install, and start a fresh install under the new user. See if it works then.

**Comment by dreashola** [comment]
> I have Windows 10, PostgreSQL 18.2, Python 3.13, PYTHOHHOME = "C:\Users\Marc\AppData\Local\Programs\Python\Python313", PYTHONPATH = "C:\Users\Marc\AppData\Local\Programs\Python\Python313" in System Env Var and the same in PATH both in System and User Env Var. Perhaps PYTHONPATH is an overkill, but I have tried with and without. C:/Program Files/PostgreSQL/18/lib/plpython3.dll is present - can be seen in File Explorer.
> 
> CREATE EXTENSION plpython3u; I get [58P01] ERROR: could not load library "C:/Program Files/PostgreSQL/18/lib/plpython3.dll": The specified module could not be found.
> 
> Where: SQL statement "CREATE FUNCTION plpython3_call_handler() RETURNS language_handler LANGUAGE c AS '$libdir/plpython3'" extension script file "plpython3u--1.0.sql", near line 3
> 
> Any idea?

I have the same issue

**Comment by marcodalo** [comment]
I wonder whether _anybody_ has plpython3u running with this configuration? EDB should know, I assume. Have they tested their release?

**Comment by anthonydb** [comment]
Hi, apologies for being away from this thread for a few months.

This morning I did a full reinstall of PostgreSQL 18.4 on Windows 11 using the EDB installer. I'm able to run `CREATE EXTENSION plpython3u;` with success.

* I have installed Python 3.13.12 in the directory `C:\Python` to make life easy.
* Under `System Variables`, I have that path in both `PATH` and `PYTHONHOME`.

I've followed these steps for years, even before EDB discontinued its Language Pack, and it has worked reliably for me. Took me a while to figure it out though.

**Comment by marcodalo** [comment]
Hi Anthony,

I got this far also, but after creating the plpythpn3u extension, have you tried creating a working python-based function in Postgresql?

Best regards,

Marc

**Comment by anthonydb** [comment]
@marcodalo 
Yes, all the code in Listings 17-16 through 17-18 for this book run fine on my Windows install:

https://github.com/anthonydb/practical-sql-2/blob/main/Chapter_17/Chapter_17.sql#L202

**Comment by marcodalo** [comment]
Well, I’ll give it another try!

Regards, Marc

**Comment by marcodalo** [comment]
Hi Anthony,

 

On a virgin Windows 11 PC I installed Postgres 18.4, Python 3.13.12 , I updated the PATH and PYTHONPATH environment variables. In PgAdmin4 I tried to create the plpython3u extension, and this is what I got:

 

could not load library "C:/Program Files/PostgreSQL/18/lib/plpython3.dll": The specified module could not be found.

 

CONTEXT: SQL statement "CREATE FUNCTION plpython3_call_handler() RETURNS language_handler

LANGUAGE c AS '$libdir/plpython3'"

extension script file "plpython3u--1.0.sql", near line 3

 

I checked in Explorer and C:/Program Files/PostgreSQL/18/lib/plpython3.dll does exist.

 

This is getting worse, not better!

 

Thank you for hanging in here !

Marc

**Comment by anthonydb** [comment]
@marcodalo 
Glad to help. One thing -- since you're replying via email would you mind cleaning up the emails before you send a reply so this thread is less cluttered? e.g., remove past comments and just send the new comment. Extra points if you edit your posts in here to declutter them! Makes it easier for others to follow.

On the latest try: I assume you 1) restarted your PC after changing the PATH and PYTHONHOME variables and 2) you were able to run `CREATE EXTENSION plpython3u;` successfully?

**Comment by marcodalo** [comment]
Yes, I did restart, and no, I could not create the plpython3u extention. The create command produced the error.

Sorry for the e-mail mess.

Regards, Marc
